### PR TITLE
Add GET /api/books/search endpoint

### PR DESCRIPTION
**Pull request by Dimagaa**
Replace the existing BookRepository with the one that will use JpaRepository interface
Add Liquibase support
Change spring.jpa.hibernate.ddl-auto property value to validate
Implement soft delete concept for the Book entity
Add missing endpoints and methods for BookService

### CODE REVIEW

**Review by ConfusedSatlan — APPROVED** [review]
Good

File: src/main/java/com/app/onlinebookstore/model/Book.java [hunk 1]
```diff
@@ -7,10 +7,17 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
 
-@SuppressWarnings("unused")
+@Getter
+@Setter
 @Entity
-@Table(name = "book")
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/model/Book.java]:
```suggestion
@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id = ?")
```

File: src/main/java/com/app/onlinebookstore/model/Book.java [hunk 2]
```diff
@@ -7,10 +7,17 @@
...
+@Where(clause = "is_deleted=false")
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/model/Book.java]:
```suggestion
@Where(clause = "is_deleted = false")
```

File: src/main/java/com/app/onlinebookstore/model/Book.java [hunk 3]
```diff
@@ -7,10 +7,17 @@
...
+@Table(name = "books")
 public class Book {
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/model/Book.java:19]: It's recommended to override equals and hashcode for entities too.

File: src/main/java/com/app/onlinebookstore/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -38,4 +42,25 @@ public BookDto getById(Long id) {
                 .map(bookMapper::toDto)
                 .orElseThrow(() -> new EntityNotFoundException("Can't find book with id: " + id));
     }
+
+    @Override
+    public BookDto updateById(Long id, CreateBookRequestDto bookRequestDto) {
+        Book book = bookMapper.toModel(bookRequestDto);
+        book.setId(id);
+        return bookMapper.toDto(bookRepository.save(book));
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/BookServiceImpl.java:55]: What if book by this id does not exist?

**Review by vovasalyha — CHANGES_REQUESTED** [review]
GJ. Left few minor suggestions.

File: src/main/java/com/app/onlinebookstore/service/impl/BookServiceImpl.java (hunk 4 above)
Comment (Dimagaa) [reply src/main/java/com/app/onlinebookstore/service/impl/BookServiceImpl.java:55]: Good point!

File: src/main/java/com/app/onlinebookstore/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -38,4 +42,28 @@ public BookDto getById(Long id) {
                 .map(bookMapper::toDto)
                 .orElseThrow(() -> new EntityNotFoundException("Can't find book with id: " + id));
     }
+
+    @Override
+    public BookDto updateById(Long id, CreateBookRequestDto bookRequestDto) {
+        if (bookRepository.findById(id).isEmpty()) {
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/BookServiceImpl.java]: Yep, something like that, but more common approach is to fetch an entity and set new fields for this entity and then save it (or not save at all if you want to use hibernate's dirty checking feature, but in your case you are returning the updated entity, so it is better to explicitly save it to return actual data)

**Review by vovasalyha — APPROVED** [review]
Approving, but please analyze my comment about update method.

File: src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+public class BookSpecificationBuilder implements SpecificationBuilder<Book, BookSearchParameters> {
+    private static final String AUTHOR_KEY = "author";
+    private static final String TITLE_KEY = "title";
+    private static final String ISBN_KEY = "isbn";
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> specification = Specification.where(null);
+        if (searchParameters.author() != null && searchParameters.author().length > 0) {
```
Comment (andrii-hoienko) [inline src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java:22]: Please use already existing method like StringUtils.isEmpty()

File: src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            specification = specification.and(bookSpecificationProviderManager
+                    .getSpecificationParameter(AUTHOR_KEY)
+                    .getSpecification(searchParameters.author()));
+        }
+        if (searchParameters.isbn() != null && searchParameters.isbn().length > 0) {
+            specification = specification.and(bookSpecificationProviderManager
+                    .getSpecificationParameter(ISBN_KEY)
+                    .getSpecification(searchParameters.isbn()));
+        }
+        if (searchParameters.title() != null && !searchParameters.title().isEmpty()) {
```
Comment (andrii-hoienko) [inline src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java:32]: Please use already existing method like ArrayUtils.isNotEmpty()

File: src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> specification = Specification.where(null);
+        if (searchParameters.author() != null && searchParameters.author().length > 0) {
+            specification = specification.and(bookSpecificationProviderManager
+                    .getSpecificationParameter(AUTHOR_KEY)
+                    .getSpecification(searchParameters.author()));
+        }
+        if (searchParameters.isbn() != null && searchParameters.isbn().length > 0) {
+            specification = specification.and(bookSpecificationProviderManager
```
Comment (andrii-hoienko) [inline src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java:28]: Blocks of code that add parameter to specification are very similar, maybe it is possible to extract common code to method?

File: src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book, BookSearchParameters> {
+    private static final String AUTHOR_KEY = "author";
+    private static final String TITLE_KEY = "title";
+    private static final String ISBN_KEY = "isbn";
+    private final SpecificationProviderManager<Book> bookSpecificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
```
Comment (andrii-hoienko) [inline src/main/java/com/app/onlinebookstore/repository/book/BookSpecificationBuilder.java:20]: OPTIONAL: You can implement strategy here

File: src/main/java/com/app/onlinebookstore/model/Book.java [hunk 10]
```diff
@@ -7,10 +7,15 @@
...
 import java.math.BigDecimal;
+import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
 
-@SuppressWarnings("unused")
+@Data
 @Entity
-@Table(name = "book")
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id = ?")
```
Comment (andrii-hoienko) [inline src/main/java/com/app/onlinebookstore/model/Book.java:16]: Why old changes displayed here?

**Review by andrii-hoienko — COMMENTED** [review]
GJ! Let`s improve your solution

**Pull request merged**